### PR TITLE
[IA-2253] Delete K8s clusters with no apps after grace period

### PR DESCRIPTION
**Pull request by kyuksel**
[IA-2033](https://broadworkbench.atlassian.net/browse/IA-2033)
[Companion Leo PR
](https://github.com/DataBiosphere/leonardo/pull/1638)

The ResourceValidator changes (soon to be its own project `janitor`™️ ) are to
- update the DB query to correctly detect K8s clusters where no apps have run for grace period (currently set to `1 hour`)
- publish a `DeleteKubernetesClusterMessage` for back Leo to process

Testing:
- Manually tested locally against dev Leo
- Unit tests being added...

### CODE REVIEW

File: project/Dependencies.scala
Comment (Qi77Qi) [inline project/Dependencies.scala:12]: I don't think we need this

File: core/src/main/scala/com/broadinstitute/dsp/models.scala
Comment (Qi77Qi) [inline core/src/main/scala/com/broadinstitute/dsp/models.scala:35]: I don't think we need this

File: core/src/main/scala/com/broadinstitute/dsp/models.scala
Comment (Qi77Qi) [inline core/src/main/scala/com/broadinstitute/dsp/models.scala:40]: I don't think we need this

File: resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala
Comment (rtitle) [inline resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:76]: I'm wondering if we could remove these LEFT JOINs and boil the interval checks into the `NOT EXISTS` subquery…

Comment (rtitle) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:76]:
```
SELECT kc.id, kc.googleProject
FROM KUBERNETES_CLUSTER kc
WHERE
  kc.status != "DELETED" AND
  NOT EXISTS (
    SELECT *
    FROM NODEPOOL n
    LEFT JOIN APP a ON n.id = a.nodepoolId
    WHERE kc.id = n.clusterId AND 
      ((n.isDefault = 0 AND a.status != "DELETED" AND a.status != "ERROR") OR
        (app.destroyedDate > now() - INTERVAL 1 HOUR) OR
        (app.dateAccessed > now() INTERVAL 1 HOUR)));
```

Comment (rtitle) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:76]: Does that work? This way the subquery checks for apps that are active OR have been recently deleted/accessed

File: resource-validator/src/main/resources/reference.conf
Comment (rtitle) [reply resource-validator/src/main/resources/reference.conf:12]: We will need to template this - I think a dedicated EV makes sense instead of parsing user creds

**Comment by rtitle** [comment]
Overall looks good, 1 comment about simplifying the SQL query

File: core/src/main/scala/com/broadinstitute/dsp/models.scala
Comment (kyuksel) [reply core/src/main/scala/com/broadinstitute/dsp/models.scala]: Huh...I thought IntelliJ implicit hints had pointed out those implicits were depended on by `Stream` piping 🤷‍♂️  Thanks for the catch.

File: resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala
Comment (kyuksel) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:76]: Great idea! I think it'll work. I just want to test it a bit more. My dev DB got into a weird state that I need to fix first.

Comment (kyuksel) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:76]: Slightly modified version of that query (note the updates within the inner `WHERE` clause below) seems to work.

```sql
            SELECT kc.id, kc.googleProject
            FROM KUBERNETES_CLUSTER kc
            WHERE
              kc.status != "DELETED" AND
              NOT EXISTS (
                SELECT *
                FROM NODEPOOL np
                LEFT JOIN APP a ON np.id = a.nodepoolId
                WHERE
                  kc.id = np.clusterId AND np.isDefault = 0 AND
                  (
                    (a.status != "DELETED" AND a.status != "ERROR") OR
                    (a.status = "DELETED" AND a.destroyedDate > now() - INTERVAL 1 HOUR) OR
                    (a.status = "ERROR" AND a.dateAccessed > now() - INTERVAL 1 HOUR)
                  ));
```

Also want to do unit testing...

**Comment by kyuksel** [comment]
@Qi77Qi @rtitle Thank you for your feedback! I think I've addressed them. Would you be able to re-review while I'm adding unit tests?

File: resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/ResourceValidator.scala
Comment (Qi77Qi) [inline resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/ResourceValidator.scala:62]: this needs to be reverted before merging

**Review by Qi77Qi — COMMENTED** [review]
lgtm so far 👍 assuming you're still working on unit tests

File: resource-validator/src/main/resources/reference.conf
Comment (Qi77Qi) [reply resource-validator/src/main/resources/reference.conf:12]: this needs to be address before merging

File: resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala
Comment (kyuksel) [inline resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:62]: @rtitle  ^^^ seems to take care of the exclusion of clusters with batch-pre-created nodepools. Added a [unit test](https://github.com/broadinstitute/leonardo-cron-jobs/pull/15/files#diff-d4eec159732c276b41661daac0de09192e20d7d261664c25dbc61d77bf0bb058R216) for that scenario as well.

File: resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala
Comment (Qi77Qi) [inline resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:61]: wouldn't `a.status = "ERROR"` alone suffice? (without checkint `dateAccessed`)

Comment (rtitle) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:61]: The `dateAccessed` check gives a grace period I think -- so we won't delete the cluster if an Error'd app is <1 hour old. 

I think I like this behavior, as a user might want to quickly recreate if they get a creation error.

**Comment by rtitle** [comment]
The SQL looks good to me; I didn't review all the Scala in detail.

File: resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala
Comment (Qi77Qi) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:61]: > we won't delete the cluster if an Error'd app is <1 hour old

If we're trying to check if the cluster is < 1hr old, shouldn't we be checking `createdDate`? still not sure why `dateAccessed` is needed

Comment (kyuksel) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:61]: Not the cluster but the app. Imagine a scenario where a user attempts to create an app but it errors out. In that case, we don't want to remove the cluster right away and instead, give the user some grace period in case they want to delete their app, create another app, etc.

Comment (kyuksel) [reply resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:61]: Hmmm...on another look, we don't seem to update `app.dateAccessed` upon app creation erroring out. In that case this query would be effectively checking `createdDate`.

File: resource-validator/src/test/scala/com/broadinstitute/dsp/resourceValidator/mocks.scala
Comment (Qi77Qi) [inline resource-validator/src/test/scala/com/broadinstitute/dsp/resourceValidator/mocks.scala:24]: not necessarily blocking this PR..but can you add this to wb-libs

Comment (kyuksel) [reply resource-validator/src/test/scala/com/broadinstitute/dsp/resourceValidator/mocks.scala:24]: yea - I can add as a follow-up PR - there's something else I'd like to add to wb-libs as well

File: core/src/main/scala/com/broadinstitute/dsp/models.scala
Comment (Qi77Qi) [inline core/src/main/scala/com/broadinstitute/dsp/models.scala:37]: minor, can you make this `TraceId` non optional. also I would probably move this to `KubernetesClusterRemover` file instead of puttng it in shared model file

**Review by Qi77Qi — APPROVED** [review]

**Review by rtitle — APPROVED** [review]

**Comment by rtitle** [comment]
Btw, I ran this in prod this morning after yesterday's release:
```
MySQL [leonardo]> SELECT kc.id, kc.googleProject, kc.creator
    ->             FROM KUBERNETES_CLUSTER kc
    ->             WHERE
    ->               kc.status != "DELETED" AND
    ->               NOT EXISTS (
    ->                 SELECT *
    ->                 FROM NODEPOOL np
    ->                 LEFT JOIN APP a ON np.id = a.nodepoolId
    ->                 WHERE
    ->                   kc.id = np.clusterId AND np.isDefault = 0 AND
    ->                   (
    ->                     (a.status != "DELETED" AND a.status != "ERROR") OR
    ->                     (a.status = "DELETED" AND a.destroyedDate > now() - INTERVAL 1 HOUR) OR
    ->                     (a.status = "ERROR" AND a.createdDate > now() - INTERVAL 1 HOUR) OR
    ->                     (a.id IS NULL)
    ->                   )
    ->                 );
+----+------------------------+--------------------------+
| id | googleProject          | creator                  |
+----+------------------------+--------------------------+
|  2 | mschatz-jamboree       | michael.schatz@gmail.com |
|  3 | anvil-tool-development | michael.schatz@gmail.com |
|  5 | ea-anvil               | afgane@gmail.com         |
+----+------------------------+--------------------------+
3 rows in set (0.054 sec)
```

**Pull request merged**
